contracts-bedrock/contracts/L1/ResourceMetering.sol

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.10;
+
+import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
+import { SignedMath } from "@openzeppelin/contracts/utils/math/SignedMath.sol";
+import { SlotPacking128x64x64 } from "../libraries/SlotPacking128x64x64.sol";
+import { Burn } from "../libraries/Burn.sol";
+
+/**
+ * @title ResourceMetering
+ * @notice ResourceMetering implements an EIP-1559 style resource metering system where pricing
+ * updates automatically based on current demand.
+ */
+contract ResourceMetering {
+    /**
+     * Along with the resource limit, determines the target resource limit.
+     */
+    int256 public constant ELASTICITY_MULTIPLIER = 2;
+
+    /**
+     * Denominator that determines max change on fee per block.
+     */
+    int256 public constant BASE_FEE_MAX_CHANGE_DENOMINATOR = 8;
+
+    /**
+     * Maximum amount of deposit gas that can be used within this block.
+     */
+    int256 public constant MAX_RESOURCE_LIMIT = 30000000;
+
+    /**
+     * Target amount of deposit gas that should be used within this block.
+     */
+    int256 public constant TARGET_RESOURCE_LIMIT = MAX_RESOURCE_LIMIT / ELASTICITY_MULTIPLIER;
+
+    /**
+     * Minimum base fee value, cannot go lower than this.
+     */
+    int256 public constant MINIMUM_BASE_FEE = 10000;
+
+    /**
+     * Initial base fee value.
+     */
+    uint128 public constant INITIAL_BASE_FEE = 1000000000;
+
+    /**
+     * EIP-1559 style gas parameters packed as follows:
+     * 128 bits: prev base fee
+     * 64 bits:  prev bought gas
+     * 64 bits:  prev block num
+     */
+    bytes32 internal resources;
+
+    /**
+     * Sets the initial resource values.
+     */
+    constructor() {
+        resources = SlotPacking128x64x64.set(INITIAL_BASE_FEE, 0, uint64(block.number));
+    }
+
+    /**
+     * Meters access to a function based an amount of a requested resource.
+     *
+     * @param _amount Amount of the resource requested.
+     */
+    modifier metered(uint64 _amount) {
+        // Record initial gas amount so we can refund for it later.
+        uint256 initialGas = gasleft();
+
+        // Run the underlying function.
+        _;
+
+        // Retrieve resource parameters.
+        (uint128 prevBaseFee, uint64 prevBoughtGas, uint64 prevBlockNum) = getResourceParams();
```

Comment:
This wrapper function is nice, but feel inconsistent with the use of `SlotPacking128x64x64.set()` above. 
I think a `setResourceParams()` function would be nice to have too.